File: flutter_theoplayer_sdk/flutter_theoplayer_sdk_android/android/src/main/kotlin/com/theoplayer/flutter/transformers/AdsTransformer.kt
package com.theoplayer.flutter.transformers

import com.theoplayer.android.api.ads.CompanionAd
import com.theoplayer.android.api.event.ads.AdIntegrationKind
import com.theoplayer.flutter.pigeon.Ad
import com.theoplayer.flutter.pigeon.AdBreak
import com.theoplayer.flutter.pigeon.IntegrationKind

typealias FlutterAd = Ad
typealias FlutterAdBreak = AdBreak
typealias FlutterCompanionAd = com.theoplayer.flutter.pigeon.CompanionAd

object AdsTransformer {

    fun toFlutterAd(ad: com.theoplayer.android.api.ads.Ad?): FlutterAd? {
        if (ad == null) {
            return null
        }

        return FlutterAd(
            id = ad.id,
            companions = ad.companions.map { toFlutterCompaionAd(it) },
            adBreak = ad.adBreak?.let { toFlutterAdBreak(it) },
            skipOffset = ad.skipOffset.toLong(),
            integration = toFlutterIntegrationKind(ad.integration),
            customIntegration = ad.customIntegration,
            customData = ad.customData,
        )
    }

    fun toFlutterAdBreak(adBreak: com.theoplayer.android.api.ads.AdBreak): FlutterAdBreak {
       return FlutterAdBreak(
           ads = adBreak.ads.map { toFlutterAd(it) },
           maxDuration = adBreak.maxDuration.toLong(),
           maxRemainingDuration = adBreak.maxRemainingDuration.toLong(),
           integration = toFlutterIntegrationKind(adBreak.integration),
           timeOffset = adBreak.timeOffset.toLong(),
           customIntegration = adBreak.customIntegration,
           customData = adBreak.customData
       )
    }

    fun toFlutterIntegrationKind(integration: AdIntegrationKind): IntegrationKind {
        return when (integration) {
            AdIntegrationKind.THEO_ADS -> IntegrationKind.THEOADS
            AdIntegrationKind.GOOGLE_IMA -> IntegrationKind.GOOGLE_IMA
            AdIntegrationKind.GOOGLE_DAI -> IntegrationKind.GOOGLE_DAI
            AdIntegrationKind.MEDIATAILOR -> IntegrationKind.MEDIATAILOR
            AdIntegrationKind.CUSTOM -> IntegrationKind.CUSTOM
        }
    }
    
    fun toFlutterCompaionAd(companionAd: CompanionAd): FlutterCompanionAd {
        return FlutterCompanionAd(
            adSlotId = companionAd.adSlotId,
            altText = companionAd.altText,
            clickThrough = companionAd.clickThrough,
            height = companionAd.height.toLong(),
            width = companionAd.width.toLong(),
            resourceURI = companionAd.resourceURI,
            type = companionAd.type
        )
    }

}

File: flutter_theoplayer_sdk/flutter_theoplayer_sdk_ios/ios/Classes/transformers/AdsTransformer.swift
//
//  AdsTransformer.swift
//  theoplayer_ios
//
//  Created by Dallos, Daniel on 15/01/2025.
//

import Foundation
import THEOplayerSDK

struct AdsTransformer {

    static func toFlutterAd(ad: THEOplayerSDK.Ad?) -> Ad? {
        guard let ad = ad else {
            return nil
        }
        
        return Ad(
            id: ad.id ?? "",
            companions: ad.companions.map({ companion in
                toFlutterCompanionAd(companionAd: companion)
            }),
            skipOffset: Int64(ad.skipOffset ?? 0),
            integration: toFlutterIntegrationKind(integration: ad.integration))
    }

    static func toFlutterAdBreak(adBreak: THEOplayerSDK.AdBreak?) -> AdBreak? {
        guard let adBreak = adBreak else {
            return nil
        }
        
        return AdBreak(
            ads: adBreak.ads.map({ ad in
                toFlutterAd(ad: ad)
            }),
            maxDuration: Int64(adBreak.maxDuration),
            maxRemainingDuration: Int64(adBreak.maxRemainingDuration),
            timeOffset: Int64(adBreak.timeOffset),
            integration: toFlutterIntegrationKind(integration: adBreak.integration)
        )
    }

    static func toFlutterIntegrationKind(integration: THEOplayerSDK.AdIntegrationKind) -> IntegrationKind {
        switch integration {
        case .theoads:
            return .theoads
        case .google_ima:
            return .googleIma
        case .google_dai:
            return .googleDai
        //case .mediaTailor:
        //    return .mediaTailor
        case .custom:
            return .custom
        @unknown default:
            fatalError("Unhandled AdIntegrationKind: \(integration)")
        }
    }

    static func toFlutterCompanionAd(companionAd: THEOplayerSDK.CompanionAd) -> CompanionAd {
        return CompanionAd(adSlotId: companionAd.adSlotId ?? "", altText: companionAd.altText ?? "", clickThrough: companionAd.clickThrough ?? "", height: Int64(companionAd.height), width: Int64(companionAd.width), resourceURI: companionAd.resourceURI!, type: companionAd.type)
    }
}
